client/beefy/src/lib.rs

```diff
@@ -346,6 +346,12 @@ where
 	R: ProvideRuntimeApi<B>,
 	R::Api: BeefyApi<B>,
 {
+	let beefy_genesis = runtime
+		.runtime_api()
+		.beefy_genesis(&BlockId::hash(best_grandpa.hash()))
+		.ok()
+		.flatten()
+		.ok_or_else(|| ClientError::Backend("BEEFY pallet expected to be active.".into()))?;
 	// Walk back the imported blocks and initialize voter either, at the last block with
 	// a BEEFY justification, or at pallet genesis block; voter will resume from there.
 	let blockchain = backend.blockchain();
```

Comment:
Sorry, unrelated to this PR - just caught my eye. Can you, please, verify me here: we're going from some header towards the genesis and eventually we may call the `expect_validator_set()?` in couple of places below. It calls the runtime API inside and if state is discarded, then it'll fail => `initialize_voter_state` will return an error => BEEFY gadget won't start. I believe I have mistake somewhere in my mind. Can you, please, verify?

Reply:
You are right, state pruning could be a problem and there is an [explicit check and error thrown for it](https://github.com/acatangiu/substrate/blob/c4b6495eb74c15ebc9f97f7c3fcd53f5911706a0/client/beefy/src/lib.rs#L413-L424) in this function a bit further in the code.

If BEEFY gadget is properly deployed (gadget already running when client imports BEEFY genesis block), this should never happen. Once gadget is running, it is persisting its state and on restarts will resume from there instead of running this initialization code so there'll be no state pruning issue.

This _could_ happen however, if the node is synced past beefy-genesis without `--beefy`, and then trying to enable `--beefy` later when state at beefy-genesis block has been pruned. In this case we just throw this error, and require resyncing or starting from some good snapshot that also includes beefy state. Nothing else we can do afaict.